File: scripts/solar/solar-common/util/BaseUtil.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__New = ____lualib.__TS__New
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["7"] = 1,["8"] = 1,["9"] = 3,["10"] = 4,["11"] = 6,["12"] = 6,["13"] = 6,["15"] = 6,["16"] = 13,["17"] = 14,["18"] = 15,["19"] = 16,["22"] = 18,["23"] = 19,["24"] = 13,["25"] = 27,["27"] = 28,["28"] = 28,["29"] = 29,["30"] = 30,["33"] = 33,["34"] = 34,["35"] = 28,["38"] = 27,["39"] = 47,["40"] = 48,["41"] = 49,["42"] = 50,["43"] = 51,["44"] = 52,["45"] = 53,["46"] = 54,["48"] = 50,["49"] = 57,["50"] = 58,["51"] = 47,["52"] = 68,["53"] = 68,["54"] = 68,["56"] = 68,["57"] = 68,["59"] = 69,["60"] = 70,["61"] = 71,["63"] = 73,["64"] = 74,["65"] = 75,["66"] = 76,["67"] = 77,["68"] = 78,["70"] = 80,["72"] = 82,["74"] = 85,["75"] = 85,["76"] = 85,["77"] = 86,["78"] = 87,["79"] = 85,["80"] = 85,["81"] = 89,["82"] = 68,["83"] = 103,["84"] = 104,["85"] = 105,["86"] = 106,["87"] = 107,["88"] = 103,["89"] = 111,["90"] = 112,["91"] = 111,["92"] = 115,["93"] = 117,["94"] = 118,["95"] = 115,["96"] = 8,["97"] = 95,["98"] = 100});
local ____exports = {}
local ____STimer = require("solar.solar-common.tool.STimer")
local STimer = ____STimer.default
time = 0
_g_time = 0
____exports.default = __TS__Class()
local BaseUtil = ____exports.default
BaseUtil.name = "BaseUtil"
function BaseUtil.prototype.____constructor(self)
end
function BaseUtil.getUUID()
    if isAsync then
        log.errorWithTraceBack("不能异步获取uuid。异步请使用AsyncUtil。getUUIDAsync()!")
        return nil
    end
    local ____exports_default_0, ____uuid_1 = ____exports.default, "uuid"
    ____exports_default_0[____uuid_1] = ____exports_default_0[____uuid_1] + 1
    return ____exports.default.uuid
end
function BaseUtil.SForGroup(whichGroup, callback)
    do
        local i = 0
        while i <= 1000000 do
            local unitHandle = FirstOfGroup(whichGroup)
            if not IsHandle(unitHandle) then
                break
            end
            callback(nil, unitHandle)
            GroupRemoveUnit(whichGroup, unitHandle)
            i = i + 1
        end
    end
end
function BaseUtil.onTimer(timeOut, onTimerCall)
    local sTimer = __TS__New(STimer)
    local count = 0
    local function callBack()
        count = count + 1
        local ____repeat = onTimerCall(nil, count)
        if not ____repeat then
            sTimer:destroy()
        end
    end
    sTimer:start(timeOut, callBack, true)
    return sTimer
end
function BaseUtil.runLater(timeOut, callBack, loopCount, immediately)
    if loopCount == nil then
        loopCount = 1
    end
    if immediately == nil then
        immediately = false
    end
    local count = 0
    if loopCount < 1 then
        return nil
    end
    local sTimer = __TS__New(STimer)
    if immediately then
        count = count + 1
        callBack(count, loopCount)
        if loopCount == 1 then
            return sTimer
        end
        sTimer:setLoopCount(loopCount - 1)
    else
        sTimer:setLoopCount(loopCount)
    end
    sTimer:start(
        timeOut,
        function()
            count = count + 1
            callBack(count, loopCount)
        end
    )
    return sTimer
end
function BaseUtil._sl_OnTime()
    _g_time = _g_time + 10
    time = _g_time
    ____exports.default.time = _g_time
    STimer._sl_OnTime()
end
function BaseUtil.getGameTime()
    return _g_time / 1000
end
function BaseUtil.init()
    ____exports.default.centerTimer = CreateTimer()
    TimerStart(____exports.default.centerTimer, 0.01, true, ____exports.default._sl_OnTime)
end
BaseUtil.uuid = 1
BaseUtil.centerTimer = nil
BaseUtil.time = 0
return ____exports

File: src/solar/solar-common/util/BaseUtil.ts
import STimer from "@/STimer";

time = 0;
_g_time = 0;
/** @noSelf **/
export default class BaseUtil {

    private static uuid: number = 1

    /**
     * 获取唯一id
     */
    static getUUID(): number {
        if (isAsync) {
            log.errorWithTraceBack("不能异步获取uuid。异步请使用AsyncUtil。getUUIDAsync()!")
            return null;
        }
        BaseUtil.uuid++;
        return BaseUtil.uuid
    }

    /**
     *遍历单位组
     *遍历完后会移除单位组的所有单位
     * @deprecated see GroupUtil.for
     */
    static SForGroup(whichGroup: group, callback: (unit: unit) => void) {
        for (let i = 0; i <= 1000000; i++) {
            let unitHandle = FirstOfGroup(whichGroup)
            if (!IsHandle(unitHandle)) {
                break
            }
            callback(unitHandle)
            GroupRemoveUnit(whichGroup, unitHandle)
        }
    }

    /**
     * 中心计时器执行回调函数
     * @param timeOut 延迟的时间 （单位：秒 精度为小数点后2位 0.01）
     * callBack 返回true 代表继续执行 返回false 代表不再执行
     * (count 为执行次数)
     * 常用质数： 3 5 7 9 (11,2)
     * (2 3 5 7 11 13 17 19 23 29 31 37 41 43 47 53 59 61 67 71 73 79 83 89 97)
     * @param onTimerCall
     */
    static onTimer(timeOut: number, onTimerCall: (count: number) => boolean): STimer {
        let sTimer = new STimer();
        let count = 0;
        let callBack = function (this: void) {
            count++;
            let repeat = onTimerCall(count);
            if (!repeat) {
                sTimer.destroy();
            }
        }
        sTimer.start(timeOut, callBack, true)
        return sTimer;
    }

    /**
     *
     * @param timeOut 延迟的时间 （单位：秒）
     * @param callBack
     * @param loopCount
     * @param immediately 立即执行一次 (示例 释放技能后 每秒造成x伤害 持续5秒  这个时候 就可以先立即执行一次 而不用等1秒才开始执行)
     */
    static runLater(timeOut: number, callBack: (this: void, count: number, maxCount: number) => void, loopCount: number = 1, immediately = false): STimer {
        let count = 0;
        if (loopCount < 1) {
            return null;
        }
        let sTimer = new STimer();
        if (immediately) {
            count++;
            callBack(count, loopCount);
            if (loopCount == 1) {
                return sTimer;
            }
            sTimer.setLoopCount(loopCount - 1);
        } else {
            sTimer.setLoopCount(loopCount);
        }

        sTimer.start(timeOut, () => {
            count++;
            callBack(count, loopCount)
        })
        return sTimer;
    }

    /**
     *  基础中心计时器
     */
    static centerTimer: timer = null;
    //毫秒计算
    /**
     * @deprecated see 全局变量 _g_time
     */
    static time = 0;


    private static _sl_OnTime(this: void) {
        _g_time += 10;
        time = _g_time;
        BaseUtil.time = _g_time;
        STimer._sl_OnTime()
    }

    //返回秒
    static getGameTime(): number {
        return _g_time / 1000;
    }

    static init() {
        //不推荐的函数 在必须使用时 可以使用 （少量使用下不会出现问题）
        BaseUtil.centerTimer = CreateTimer();
        TimerStart(BaseUtil.centerTimer, 0.01, true, BaseUtil._sl_OnTime);
    }

}

